sklearn/tree/_criterion.pyx

```diff
@@ -79,22 +79,22 @@ cdef class Criterion:
 
         pass
 
-    cdef void reset(self) nogil:
+    cdef int reset(self) nogil except -1:
```

Comment:
The problem with that is `void ... except *` has a tiny overhead that gets amplified as the functions are called multiple times. `in ... except -1` has very low overhead as the `PyErr_Occurred()` is not called to check for any Python errors that were raised...